pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -1242,39 +1244,146 @@ func (vs *VSphere) CreateVolume(volumeOptions *VolumeOptions) (volumePath string
-			DiskType:    diskFormat,
+	virtualDiskSpec := &types.VirtualDisk{
+		CapacityInKB: 4000000,
+	}
+	virtualDeviceSpec := types.VirtualDevice{
+		Key:           0,
+		ControllerKey: 1000,
+		UnitNumber:    new(int32), // zero default value
+	}
+
+	backingObjectSpec := &types.VirtualDiskFlatVer2BackingInfo{
+		DiskMode: string(types.VirtualDiskModePersistent),
+		VirtualDeviceFileBackingInfo: types.VirtualDeviceFileBackingInfo{
+			FileName: "[" + ds.Name() + "] " + volumeOptions.Name + ".vmdk",
 		},
-		CapacityKb: int64(volumeOptions.CapacityKB),
 	}
+	if diskFormat == ThinDiskType {
+		backingObjectSpec.ThinProvisioned = types.NewBool(true)
+	} else if diskFormat == EagerZeroedThickDiskType {
+		backingObjectSpec.EagerlyScrub = types.NewBool(true)
+	} else {
+		backingObjectSpec.ThinProvisioned = types.NewBool(false)
+	}
+
+	virtualDeviceSpec.Backing = backingObjectSpec
+	virtualDiskSpec.VirtualDevice = virtualDeviceSpec
+	diskConfigSpec.Device = virtualDiskSpec
 
-	// Create virtual disk
-	task, err := virtualDiskManager.CreateVirtualDisk(ctx, vmDiskPath, dc, vmDiskSpec)
+	if volumeOptions.StorageProfileData != "" {
+		storagePolicySpec := &types.VirtualMachineDefinedProfileSpec{
+			ProfileId: "",
+			ProfileData: &types.VirtualMachineProfileRawData{
+				ExtensionKey: "com.vmware.vim.sps",
+				ObjectData:   volumeOptions.StorageProfileData,
+			},
+		}
+		diskConfigSpec.Profile = append(diskConfigSpec.Profile, storagePolicySpec)
+	}
+	virtualMachineConfigSpec.DeviceChange = append(virtualMachineConfigSpec.DeviceChange, scsiDeviceConfigSpec)
+	virtualMachineConfigSpec.DeviceChange = append(virtualMachineConfigSpec.DeviceChange, diskConfigSpec)
+
+	dcFolders, err := dc.Folders(ctx)
+	vmRegex := vs.cfg.Global.WorkingDir + vs.localInstanceID
+	currentVM, err := f.VirtualMachine(ctx, vmRegex)
+	if err != nil {
+		return "", err
+	}
+
+	currentVMHost, err := currentVM.HostSystem(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	resourcePool, err := currentVMHost.ResourcePool(ctx)
 	if err != nil {
 		return "", err
 	}
+
+	task, err := dcFolders.VmFolder.CreateVM(ctx, virtualMachineConfigSpec, resourcePool, nil)
+	if err != nil {
+		return "", err
+	}
+
 	err = task.Wait(ctx)
 	if err != nil {
+		dummyVMToDelete, err := f.VirtualMachine(ctx, dummyVMName)
+		if err != nil {
+			return "", err
+		}
+		destroyTask, err := dummyVMToDelete.Destroy(ctx)
+		if err != nil {
+			return "", err
+		}
+		err = destroyTask.Wait(ctx)
+		if err != nil {
+			return "", err
+		}
+		return "", err
+	}
+
+	dummyVM, err := f.VirtualMachine(ctx, dummyVMName)
+	if err != nil {
 		return "", err
 	}
+	vmDiskPath, err := getVMDiskPath(vs.client, dummyVM, volumeOptions.Name+".vmdk")
+	if err != nil {
+		return "", fmt.Errorf("Failed to get disk path for disk on VM: %q with err: %+v", dummyVMName, err)
+	}
+	if vmDiskPath == "" {
+		return "", fmt.Errorf("Unable to find the path for the disk on VM: %q with err: %+v", dummyVMName, err)
+	}
 
+	err = vm.Unregister(ctx)
```

Comment:
This disk is the one created from dynamic policy provisioning. All disks created using storage policy will have a format "k8s_pvc_{UUID}.vmdk". Why is there a need to differentiate it. 

Yes, this will leave vmx file data behind. I think this would be very very small.